syntax = "proto3";

package colonio.api;

import "core/colonio_api.proto";
import "core/map_api.proto";
import "core/pubsub2d_api.proto";

message NoParam {
}

message Call {
  uint32 id = 1;
  uint32 channel = 2;

  oneof param {
    colonio.Connect colonio_connect = 1001;
    NoParam colonio_disconnect = 1002;
    NoParam colonio_get_local_nid = 1003;
    colonio.SetPosition colonio_set_position = 1004;

    map_api.Get map_get = 2001;
    map_api.Set map_set = 2002;

    pubsub2d.Publish pubsub2d_publish = 3001;
  }
}

message Success {
}

message Failure {
  uint32 code = 1;
  string message = 2;
}

message Reply {
  uint32 id = 1;

  oneof param {
    Success success = 11;
    Failure failure = 12;

    colonio.ConnectReply colonio_connect = 1001;
    colonio.GetLocalNIDReply colonio_get_local_nid = 1003;
    colonio.SetPositionReply colonio_set_position = 1004;

    map_api.GetReply map_get = 2001;
  }
}

message Event {
  uint32 channel = 1;

  oneof param {
    colonio.LogEvent colonio_log = 1001;
    colonio.DebugEvent colonio_debug = 1002;

    pubsub2d.OnEvent pubsub2d_on = 3001;
  }
}
